pkg/net/libp2p/libp2p_transport.go

```diff
@@ -0,0 +1,454 @@
+			continue
+		}
+		parsedAddrInfo[nodeID] = info
+	}
+
+	t.connsLock.Lock()
+	for nodeID := range parsedAddrInfo {
+		if nodeID == t.ownID {
+			continue
+		}
+		_, found := t.conns[nodeID]
+		if !found {
+			t.conns[nodeID] = &connInfo{AddrInfo: parsedAddrInfo[nodeID], Stream: nil}
+		}
+	}
+	t.connsLock.Unlock()
+
+	t.connect(ctx, maputil.GetKeys(parsedAddrInfo))
+}
+
+func (t *Transport) Send(ctx context.Context, dest types.NodeID, msg *messagepb.Message) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-t.stopChan:
+		return fmt.Errorf("transport stopped")
+	default:
+	}
+
+	out, err := t.encode(msg)
+	if err != nil {
+		return err
+	}
+
+	err = t.sendPayload(dest, out)
+	if err != nil {
+		t.connect(ctx, []types.NodeID{dest})
+		return errors.Wrapf(err, "%s failed to send data to %s", t.ownID, dest)
+	}
+
+	return nil
+}
+
+func (t *Transport) ImplementsModule() {}
+
+func (t *Transport) EventsOut() <-chan *events.EventList {
+	return t.incomingMessages
+}
+
+func (t *Transport) ApplyEvents(ctx context.Context, eventList *events.EventList) error {
+	iter := eventList.Iterator()
+	for event := iter.Next(); event != nil; event = iter.Next() {
+
+		switch e := event.Type.(type) {
+		case *eventpb.Event_Init:
+			// no actions on init
+		case *eventpb.Event_SendMessage:
+			for _, destID := range e.SendMessage.Destinations {
+				if types.NodeID(destID) == t.ownID {
+					// Send message to myself bypassing the network.
+					// The sending must be done in its own goroutine in case writing to gt.incomingMessages blocks.
+					// (Processing of input events must be non-blocking.)
+					receivedEvent := events.MessageReceived(
+						types.ModuleID(e.SendMessage.Msg.DestModule),
+						t.ownID,
+						e.SendMessage.Msg,
+					)
+					go func() {
+						select {
+						case t.incomingMessages <- events.ListOf(receivedEvent):
+						case <-ctx.Done():
+						}
+					}()
+				} else {
+					// Send message to another node.
+					if err := t.Send(ctx, types.NodeID(destID), e.SendMessage.Msg); err != nil {
+						t.logger.Log(logging.LevelError, "failed to send a message", "err", err)
+					}
+				}
+			}
+		default:
+			return fmt.Errorf("unexpected event: %T", event.Type)
+		}
+	}
+
+	return nil
+}
+
+func (t *Transport) runStreamUpdater() {
+	go func() {
+		t.logger.Log(logging.LevelDebug, "stream updater started", "src", t.ownID)
+		defer t.logger.Log(logging.LevelDebug, "stream updater stopped", "src", t.ownID)
+
+		for {
+			select {
+			case <-t.stopChan:
+				t.logger.Log(logging.LevelError, "stream updater received stop signal", "src", t.ownID)
+				return
+			case one := <-t.streamChan:
+				t.connsLock.Lock()
+				conn, found := t.conns[one.NodeID]
+				if !found {
+					t.connsLock.Unlock()
+					continue
+				}
+				conn.Stream = one.Stream
+				t.connsLock.Unlock()
+				t.logger.Log(logging.LevelDebug, "updated stream", "src", t.ownID, "nodeID", one.NodeID)
```

Comment:
If we found the stream, let's try to close the old one (after releasing the lock).

Reply:
Fixed